src/message_filter.rs

```diff
@@ -56,92 +62,58 @@ impl<Message: Hash> MessageFilter<Message> {
     pub fn insert(&mut self, message: &Message) -> usize {
         self.remove_expired();
         let hash_code = hash(message);
-        if let Some(index) = self.entries.iter().position(|t| t.hash_code == hash_code) {
-            let mut timestamped_message = self.entries.remove(index);
-            timestamped_message.update_expiry_point(self.time_to_live);
-            let count = timestamped_message.increment_count();
-            self.entries.push(timestamped_message);
-            count
-        } else {
-            self.entries.push(TimestampedMessage::new(hash_code, self.time_to_live));
-            1
+        let expiry = Instant::now() + self.time_to_live;
+        self.timeout_queue.push_back((hash_code, expiry));
+        match self.count.entry(hash_code) {
+            Entry::Occupied(entry) => {
+                let &mut (ref mut c, ref mut t) = entry.into_mut();
+                *t = expiry;
+                *c += 1;
+                *c
+            }
+            Entry::Vacant(entry) => entry.insert((1, expiry)).0,
         }
     }
 
     /// Returns the number of times this message has already been inserted.
     #[cfg(test)]
     pub fn count(&self, message: &Message) -> usize {
         let hash_code = hash(message);
-        self.entries
-            .iter()
-            .find(|t| t.hash_code == hash_code)
-            .map_or(0, |t| t.count)
+        self.count.get(&hash_code).map_or(0, |&(count, _)| count)
     }
 
     /// Removes any expired messages, then returns whether `message` exists in the filter or not.
     pub fn contains(&mut self, message: &Message) -> bool {
         self.remove_expired();
-        let hash_code = hash(message);
-        self.entries.iter().any(|entry| entry.hash_code == hash_code)
+        self.count.contains_key(&hash(message))
     }
 
     /// Remove the entry for `message`, regardless of how many times it was previously inserted.
     pub fn remove(&mut self, message: &Message) {
-        let hash_code = hash(message);
-        if let Some(index) = self.entries.iter().position(|t| t.hash_code == hash_code) {
-            let _old_val = self.entries.remove(index);
-        }
+        let _old_val = self.count.remove(&hash(message));
```

Comment:
It doesn't matter and my motivation for this PR was exactly to avoid ever having to iterate through all of them.